tiledb/sm/global_state/libcurl_state.cc

```diff
@@ -35,26 +35,57 @@
 
 #ifdef TILEDB_SERIALIZATION
 #include <curl/curl.h>
+
+// This ifdef'ed definition of do_init_ only exists
+// because the preprocessor is unable to parse
+// CURL_GLOBAL_DEFAULT properly inside a constexpr block
+// when the curl.h header hasn't been included.
+
+CURLcode do_init_() {
+  return curl_global_init(CURL_GLOBAL_DEFAULT);
+}
+
+#else
+
+int do_init_() {
+  return 0;
+}
+
 #endif
 
 using namespace tiledb::common;
 
-namespace tiledb {
-namespace sm {
-namespace global_state {
+namespace tiledb::sm::global_state {
 
-Status init_libcurl() {
-#ifdef TILEDB_SERIALIZATION
-  auto rc = curl_global_init(CURL_GLOBAL_DEFAULT);
-  if (rc != 0)
-    return LOG_STATUS(Status_Error(
-        "Cannot initialize libcurl global state: got non-zero return code " +
-        std::to_string(rc)));
-#endif
+class LibCurlState {
+  public:
+    static Status init();
+
+  private:
+    static std::once_flag init_done_;
+    static Status status_;
+};
+
+std::once_flag LibCurlState::init_done_;
+Status LibCurlState::status_;
 
-  return Status::Ok();
+Status LibCurlState::init() {
```

Comment:
This class isn't immune from C.41 concerns. Instead of using `init`, we can simply construct an instance for its side-effects. This also allows avoid a new use of `Status` (against policy).

To use it, we can make it a member variable in all the VFS classes that use curl. It should be the first-declared member variable, so that it's initialized first and all curl calls made thereafter will be on an initialized library.

With the approach, the function `init_libcurl` disappears, as does its call in `GlobalState::init`, complete remove any relationship. As such, we can move curl initialization into `filesystem` directory, where it will be used.

Reply:
I've rearranged this to follow your suggestion of having a simple class's constructor ensure that libcurl is initialized and then added member variables to each of the VFS classes as well as `tiledb::sm::Curl` in `tiledb/sm/rest/curl.h`.

I haven't yet moved the initializer class though. Given that cURL is also used in `tiledb/sm/rest/curl.{h,cc}` I hesitated on if to move it and if there was somewhere better that isn't global state or filesystem.

I'm also fairly hesitant to remove the libcurl initialization call from global_state given that I can't prove to myself that I'm not missing any other uses of curl.

Reply:
The build spec is pretty clear what the dependencies are. We only include curl in certain circumstances. From `tiledb/CMakeLists.txt`:
```
# Libcurl
if (TILEDB_S3 OR TILEDB_AZURE OR TILEDB_GCS OR TILEDB_SERIALIZATION)
```
I think we're on safe ground with the assumption that nothing else has a curl dependency. Given that, I think we ought to move all this code out of `global_state/`.

Reply:
> somewhere better that isn't global state or filesystem

`filesystem/` is fine for us. It's at the bottom of the dependency tree, so it won't pose any problems. The alternative, which seems like overkill, is to create a new directory for both `filesystem/` and `rest/` to depend upon.

I'm also thinking that `global_state/` is going to disappear as a directory itself when its functionality gets rewritten to work right.

Reply:
Ah! Excellent idea looking for the build switch there for reassurance something else hasn't snuck in. I'll move this to filesystem and call it good. I also agree that it seems likely global_state will end up being emptied out as we refactor that code.

Reply:
Probably not an actual problem. AWS calls `curl_global_init`, for example. I haven't checked the others.

It might be a different problem that it's getting multiply initialized, but if it's an actual problem, it's not the issue at present.